Security/Http/Authenticator/OAuthAuthenticator.php

```diff
@@ -144,72 +171,73 @@ public function authenticate(Request $request): Passport
             throw new AuthenticationServiceException('loadUserByOAuthUserResponse() must return a UserInterface.');
         }
 
-        $this->rawToken = $token->getRawToken();
-        $this->resourceOwnerName = $resourceOwner->getName();
-        $this->refreshToken = $token->getRefreshToken();
-        $this->createdAt = $token->getCreatedAt();
-
-        return new SelfValidatingPassport(
-            class_exists(UserBadge::class)
-                ? new UserBadge(
-                    method_exists($user, 'getUserIdentifier') ? $user->getUserIdentifier() : $user->getUsername(),
-                    static function () use ($user) { return $user; }
-                )
-                : $user
-        );
+        return $this->recreateToken($token, $user);
     }
 
     /**
-     * @param Passport|SelfValidatingPassport $passport
+     * @template T of OAuthToken
+     *
+     * @param T              $token
+     * @param ?UserInterface $user
+     *
+     * @return T
      */
-    public function createAuthenticatedToken($passport, string $firewallName): TokenInterface
+    public function recreateToken(OAuthToken $token, ?UserInterface $user = null): OAuthToken
     {
-        $token = $this->createToken($passport, $firewallName);
+        $user = $user instanceof UserInterface ? $user : $token->getUser();
+
+        $tokenKlass = \get_class($token);
+        if ($user) {
+            $newToken = new $tokenKlass(
+                $token->getRawToken(),
+                method_exists($user, 'getRoles') ? $user->getRoles() : []
+            );
+            $newToken->setUser($user);
+        } else {
+            $newToken = new $tokenKlass($token->getRawToken());
```

Comment:
```suggestion
        $tokenClass = \get_class($token);
        if ($user) {
            $newToken = new $tokenClass(
                $token->getRawToken(),
                method_exists($user, 'getRoles') ? $user->getRoles() : []
            );
            $newToken->setUser($user);
        } else {
            $newToken = new $tokenClass($token->getRawToken());
```

Or with above new method ;)